projects/driver_controls/src/ads1015.c

```diff
@@ -0,0 +1,169 @@
+#include "ads1015.h"
+
+#include <stdbool.h>
+#include <stdio.h>
+
+#include "i2c.h"
+#include "gpio_it.h"
+#include "ads1015_bitmasks.h"
+
+// Union definition to help process 16-bit registers
+// Due to endian-ness, index 0 refers to the most significant byte
+typedef union ADS1015Register {
+  uint16_t raw;
+  uint8_t data[2];
+} ADS1015Register;
+
+typedef struct ADS1015Interrupt {
+  ADS1015Callback callback;
+  void *context;
+  int16_t reading;
+} ADS1015Interrupt;
+
+static GPIOAddress s_address;
+static I2CPort s_i2c_port;
+static ADS1015Channel s_current_channel;
+
+static ADS1015Interrupt s_interrupts[NUM_ADS1015_CHANNELS];
+
+static StatusCode prv_channel_valid(ADS1015Channel channel) {
+  if (channel >= NUM_ADS1015_CHANNELS) {
+    return STATUS_CODE_INVALID_ARGS;
+  }
+
+  return STATUS_CODE_OK;
+}
+
+// Initiates a read as outlined in section 8.5.3 of the datasheet
+static StatusCode prv_read(I2CPort i2c_port, uint8_t reg, uint8_t *data) {
+  status_ok_or_return(i2c_write(i2c_port, ADS1015_I2C_ADDRESS, &reg, 1));
+  status_ok_or_return(i2c_read(i2c_port, ADS1015_I2C_ADDRESS, data, 2));
+
+  return STATUS_CODE_OK;
+}
+
+// Initiates a write as outlined in section 8.5.3 of the datasheet
+static StatusCode prv_write(I2CPort i2c_port, uint8_t reg, uint8_t *data) {
+  // For writes, the input data must be sent in the same frame as the address pointer register
+  uint8_t write_data[] = { reg, data[0], data[1] };
+  status_ok_or_return(i2c_write(i2c_port, ADS1015_I2C_ADDRESS, write_data, 3));
+
+  return STATUS_CODE_OK;
+}
+
+static void prv_interrupt_handler(GPIOAddress *address, void *context) {
+  ADS1015Register reg;
+
+  // Obtain ADC readings
+  status_ok_or_return(prv_read(s_i2c_port, ADS1015_CONVERSION_REGISTER, reg.data));
+  s_interrupts[s_current_channel].reading = (reg.data[0] << 4) | (reg.data[1] >> 4);
+
+  if (s_interrupts[s_current_channel].callback != NULL) {
+    s_interrupts[s_current_channel].callback(s_current_channel,
+                                              s_interrupts[s_current_channel].context);
+  }
+
+  // Update so that the ADC reads from the next channel
+  s_current_channel = (s_current_channel + 1) % NUM_ADS1015_CHANNELS;
+
+  // Obtain the current value for the config register
+  status_ok_or_return(prv_read(s_i2c_port, ADS1015_CONFIG_REGISTER, reg.data));
+
+  // Set bits 12-14 based on the current channel
+  switch (s_current_channel) {
```

Comment:
It's still staying in single-ended mode. It's just alternating between channels.